tree/tree/src/InternalTreeUtils.cxx

```diff
@@ -0,0 +1,179 @@
+/*************************************************************************
+ * Copyright (C) 1995-2021, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include "ROOT/InternalTreeUtils.hxx"
+#include "TTree.h"
+#include "TChain.h"
+#include "TFile.h"
+#include "TFriendElement.h"
+
+#include <utility> // std::pair
+#include <vector>
+#include <stdexcept> // std::runtime_error
+#include <string>
+
+namespace ROOT {
+namespace Internal {
+namespace TreeUtils {
+
+////////////////////////////////////////////////////////////////////////////////
+/// \fn std::vector<std::string> GetFileNamesFromTree(const TTree &tree)
+/// \ingroup tree
+/// \brief Get and store the file names associated with the input tree.
+/// \param[in] tree The tree from which friends information will be gathered.
+/// \throws std::runtime_error If no files could be associated with the input tree.
+std::vector<std::string> GetFileNamesFromTree(const TTree &tree)
+{
+   std::vector<std::string> filenames;
+
+   // If the input tree is a TChain, traverse its list of associated files.
+   if (auto chain = dynamic_cast<const TChain *>(&tree)) {
+      const TObjArray *filelist = chain->GetListOfFiles();
+      const auto nFiles = filelist->GetEntries();
+      if (nFiles == 0)
+         throw std::runtime_error("The provided chain of files is empty");
```

Comment:
```suggestion
         throw std::runtime_error("The provided chain of files is empty.");
```

Reply:
Alright thanks! If Philippe approves too I will include this modification in another commit :+1: